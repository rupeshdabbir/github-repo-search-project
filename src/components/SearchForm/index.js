import React, { useState } from 'react';
import { StyleSheet, css } from 'aphrodite/no-important';
import Dropdown from '../Dropdown/index';
import 'suitcss-utils-display';
import 'react-dropdown/style.css';
import SearchIcon from './search.svg';

const SearchForm = (props) => {
  const [initialInputValue, languageStr] = props.initialInputValue.split(' ');
  const [inputValue, setInputValue] = useState(initialInputValue);

  let dropdownValue = languageStr && languageStr.length ? languageStr.split(':')[1] : 'JavaScript';

  const handleOnChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleOnSubmit = (event) => {
    event.preventDefault();
    if (!inputValue && !dropdownValue) {
      return;
    }
    props.onSubmit({ inputValue, dropdownValue });
  };

  const onDropdownSelect = (value) => {
    dropdownValue = value;
  };

  return (
    <form className={css(styles.SearchForm)} onSubmit={handleOnSubmit}>
      <label className="u-hiddenVisually" htmlFor="searchInput">
        Search for a repo
      </label>
      <input
        className={css(styles.SearchForm_input)}
        type="text"
        id="searchInput"
        placeholder="Search for a repo"
        value={inputValue}
        onChange={handleOnChange}
      />
      <Dropdown onSelect={onDropdownSelect} default={dropdownValue} />

      <button className={css(styles.SearchForm_btn)} type="submit">
        <SearchIcon width={22} height={24} />
        <span className="u-hiddenVisually">Submit</span>
      </button>
    </form>
  );
};

const styles = StyleSheet.create({
  SearchForm: {
    position: 'relative',
  },

  SearchForm_input: {
    padding: 8,
    paddingRight: 40,
    width: '100%',
    border: '1px solid #bbb',
    boxShadow: 'inset 0 2px 2px rgba(0, 0, 0, .1);',
  },

  SearchForm_btn: {
    padding: '3px 14px',
    position: 'absolute',
    top: 4,
    right: 0,
  },
});

export default SearchForm;
