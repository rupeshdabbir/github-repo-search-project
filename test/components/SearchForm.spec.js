import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { StyleSheetTestUtils } from 'aphrodite';

import SearchForm from 'components/SearchForm';

Enzyme.configure({ adapter: new Adapter() });

describe('Component: SearchForm', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeAll(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterAll(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
    jest.clearAllMocks();
  });

  it('should render as expected', () => {
    const component = renderer.create(<SearchForm onSubmit={jest.fn()} initialInputValue="" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('when an initialInputValue prop is absent', () => {
    it('should set inputValue to an empty string', () => {
      const wrapper = shallow(<SearchForm onSubmit={jest.fn()} initialInputValue="" />);
      expect(wrapper.props().initialInputValue).toEqual(undefined);
    });
  });

  describe('when a different initialInputValue is set', () => {
    it('should update the state', () => {
      const wrapper = shallow(<SearchForm initialInputValue={'hello'} onSubmit={jest.fn()} />);
      const value = wrapper.find('#searchInput').props().value;
      expect(value).toEqual('hello');
    });
  });

  describe('when the form is submitted', () => {
    it('should pass the value to the onSubmit function', () => {
      const spy = jest.fn();
      const response = { dropdownValue: 'JavaScript', inputValue: 'angular+language:JavaScript' };
      const wrapper = shallow(<SearchForm onSubmit={spy} initialInputValue="angular+language:JavaScript" />);
      wrapper.simulate('submit', { preventDefault: (f) => f });
      expect(spy).toHaveBeenCalledWith(response);
    });
  });
});
