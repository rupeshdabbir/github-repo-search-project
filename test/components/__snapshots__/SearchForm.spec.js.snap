// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchForm should render as expected 1`] = `
<form
  className="SearchForm_e296pg"
  onSubmit={[Function]}
>
  <label
    className="u-hiddenVisually"
    htmlFor="searchInput"
  >
    Search for a repo
  </label>
  <input
    className="SearchForm_input_12zt507"
    id="searchInput"
    onChange={[Function]}
    placeholder="Search for a repo"
    type="text"
    value=""
  />
  <div
    className="Dropdown-root"
  >
    <div
      aria-haspopup="listbox"
      className="Dropdown-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="Dropdown-placeholder is-selected"
      >
        JavaScript
      </div>
      <div
        className="Dropdown-arrow-wrapper"
      >
        <span
          className="Dropdown-arrow"
        />
      </div>
    </div>
  </div>
  <button
    className="SearchForm_btn_13vgzd5"
    type="submit"
  >
    <svg />
    <span
      className="u-hiddenVisually"
    >
      Submit
    </span>
  </button>
</form>
`;
